runtime/parachains/src/paras.rs

```diff
@@ -966,18 +969,34 @@ impl<T: Config> Pallet<T> {
 			if up.is_some() {
 				T::DbWeight::get().reads_writes(1, 0)
 			} else {
+				// The block number the upgrade is expected at
+				// (calculated from the point the parachain block was build on).
 				let expected_at = relay_parent_number + cfg.validation_upgrade_delay;
 				let next_possible_upgrade_at =
 					relay_parent_number + cfg.validation_upgrade_frequency;
+				// The block number of the current block.
+				let now = frame_system::Pallet::<T>::current_block_number();
+
+				// If the expected upgrade block is in the past, we need
+				// to move this up to the current block. Aka when a parachain block
+				// will be enacted that was build on this or any descendent block the
+				// runtime upgrade should be enacted.
+				let expected_at = if expected_at <= now { now } else { expected_at };
 
 				*up = Some(expected_at);
 
-				<Self as Store>::UpcomingUpgrades::mutate(|upcoming_upgrades| {
-					let insert_idx = upcoming_upgrades
-						.binary_search_by_key(&expected_at, |&(_, b)| b)
-						.unwrap_or_else(|idx| idx);
-					upcoming_upgrades.insert(insert_idx, (id, expected_at));
-				});
+				// We set the go-ahead signal directly, if the expected upgrade block
+				// is in the past.
+				if expected_at <= now {
+					UpgradeGoAheadSignal::<T>::insert(&id, UpgradeGoAhead::GoAhead);
+				} else {
+					<Self as Store>::UpcomingUpgrades::mutate(|upcoming_upgrades| {
+						let insert_idx = upcoming_upgrades
+							.binary_search_by_key(&expected_at, |&(_, b)| b)
+							.unwrap_or_else(|idx| idx);
+						upcoming_upgrades.insert(insert_idx, (id, expected_at));
+					});
+				}
```

Comment:
The weight profile of this function was changed, so we probably should update it.